src/selector_widget.cpp

```diff
@@ -0,0 +1,433 @@
+ *
+ * @see Window::OnEditboxChanged()
+ */
+void SelectorWidget::OnEditboxChanged(WidgetID wid)
+{
+	if (wid != WID_SELECTOR_EDITBOX) {
+		return;
+	}
+	this->string_filter.SetFilterTerm(this->editbox.text.buf);
+	this->w->InvalidateData(0);
+	this->vscroll->SetCount(this->filtered_list.size());
+}
+
+/**
+ * Selector widget's DrawWidget method
+ * This function is ment to be called from the parrent's window's DrawWidget
+ * method
+ *
+ * @param widget  Widget number.
+ * @param[in,out] size Size of the widget.
+ * @param padding Recommended amount of space between the widget content and the widget edge.
+ * @param[in,out] fill Fill step of the widget.
+ * @param[in,out] resize Resize step of the widget.
+ *
+ * All of theese parameters should be passed straight from the parrent's window's DrawWidget
+ * Event handler, without any changes, or conditional checks
+ *
+ * This function calls a profile defined function this->profile.DrawSection(),
+ * that actually draws the fields
+ *
+ * @see Window::DrawWidget()
+ * @see SelectorWidget::Profile
+ */
+void SelectorWidget::DrawWidget(const Rect &r, WidgetID widget)
+{
+	if (widget != WID_SELECTOR_MATRIX) {
+		return;
+	}
+	Rect line = r.WithHeight(row_height);
+
+	const bool rtl = _current_text_dir == TD_RTL;
+
+	auto [first, last] = this->vscroll->GetVisibleRangeIterators(filtered_list);
+
+	for (auto it = first; it != last; ++it) {
+		const Rect ir = line.Shrink(WidgetDimensions::scaled.framerect);
+
+		if (!this->shown[*it]) {
+			DrawFrameRect(ir, COLOUR_BROWN, FR_LOWERED);
+		}
+		this->profile.DrawSection(this, *it, ir.Indent(WidgetDimensions::scaled.fullbevel.Horizontal(), rtl));
+
+		line = line.Translate(0, row_height);
+	}
+}
+
+/**
+ * This function is one of the possible functions ment to be used as part of the DrawWidget::Profile
+ * It draws "section" (line) of the scrollable list
+ * @param wid The current SelectionWidget object pointer
+ * @param id the "id" of the item to be drawn, can mean different things depending on the profile
+ * e.g when this widget displays companies, the id is a unique and valid CompanyID
+ * @param r the rectangle where the item is to be drawn
+ * @see SelectorWidget::DrawSection()
+ * @see SelectorWidget::DrawWidget()
+ * @see SelectorWidget::Profile
+ */
+void DrawSectionCompany(SelectorWidget *wid, int id, const Rect &r)
+{
+	const CompanyID cid = (CompanyID)id;
+	assert(Company::IsValidID(cid));
+
+	const bool rtl = _current_text_dir == TD_RTL;
+	const Dimension d = GetSpriteSize(SPR_COMPANY_ICON);
+
+	DrawCompanyIcon(cid, rtl ? r.right - d.width : r.left, CenterBounds(r.top, r.bottom, d.height));
+
+	const Rect text = r.Indent(d.width + WidgetDimensions::scaled.hsep_normal, rtl);
+
+	SetDParam(0, cid);
+	if (wid->selected_id.value_or(INVALID_OWNER) == cid) {
+		DrawString(text.left, text.right, CenterBounds(text.top, text.bottom, GetCharacterHeight(FS_NORMAL)), STR_COMPANY_NAME,  TC_WHITE);
```

Comment:
Extra space in here, but you could also determine the text colour with `TextColour colour = (wid->selected_id.value_or(INVALID_OWNER) == cid) ? TC_WHITE : TC_BLACK;` and then pass that to a single `DrawString()`.

```suggestion
		DrawString(text.left, text.right, CenterBounds(text.top, text.bottom, GetCharacterHeight(FS_NORMAL)), STR_COMPANY_NAME, TC_WHITE);
```